File: java/leetcode/binarySearch/find_smallest_letter_greater_than_target.java
/**
 * 744. Given a list of sorted characters letters containing only lowercase letters, and given a target letter target, find the smallest element in the list that is larger than the given target.

Letters also wrap around. For example, if the target is target = 'z' and letters = ['a', 'b'], the answer is 'a'.

Examples:

Input:
letters = ["c", "f", "j"]
target = "a"
Output: "c"

Input:
letters = ["c", "f", "j"]
target = "c"
Output: "f"

Input:
letters = ["c", "f", "j"]
target = "d"
Output: "f"

Input:
letters = ["c", "f", "j"]
target = "g"
Output: "j"

Input:
letters = ["c", "f", "j"]
target = "j"
Output: "c"

Input:
letters = ["c", "f", "j"]
target = "k"
Output: "c"
Note:

letters has a length in range [2, 10000].
letters consists of lowercase letters, and contains at least 2 unique letters.
target is a lowercase letter.
 */

public class find_smallest_letter_greater_than_target {

    public char nextGreatestLetter(char[] letters, char target) {
        int n = letters.length;
        int l = 0, h = n-1;
        while (l <= h) {
            int m = l + (h-l) / 2;
            if (letters[m] <= target) {
                l = m + 1;
            } else {
                h = m - 1;
            }
        }
        return l < n ? letters[l] : letters[0];
    }
}

File: golang/leetcode/binarysearch/find_smallest_letter_greater_than_target.go
package binarysearch

// 744. Find Smallest Letter Greater Than Target

func nextGreatestLetter(letters []byte, target byte) byte {
	n := len(letters)
	l, h := 0, n-1
	var m int
	for l <= h {
		m = l + (h-l)/2
		if letters[m] <= target {
			l = m + 1
		} else {
			h = m - 1
		}
	}
	if l < n {
		return letters[l]
	}
	return letters[0]
}
